src/Avalonia.Base/Media/Fonts/EmbeddedFontCollection.cs

```diff
@@ -0,0 +1,281 @@
+                {
+                    return true;
+                }
+
+                if (key.Weight != FontWeight.Normal)
+                {
+                    if (TryFindStretchFallback(glyphTypefaces, new FontCollectionKey(key.Style, FontWeight.Normal, key.Stretch), out glyphTypeface))
+                    {
+                        return true;
+                    }
+                }
+
+                key = new FontCollectionKey(key.Style, key.Weight, FontStretch.Normal);
+            }
+
+            if (TryFindWeightFallback(glyphTypefaces, key, out glyphTypeface))
+            {
+                return true;
+            }
+
+            if (TryFindStretchFallback(glyphTypefaces, key, out glyphTypeface))
+            {
+                return true;
+            }
+
+            //Take the first glyph typeface we can find.
+            foreach (var typeface in glyphTypefaces.Values)
+            {
+                glyphTypeface = typeface;
+
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool TryFindStretchFallback(
+            Dictionary<FontCollectionKey, IGlyphTypeface> glyphTypefaces,
+            FontCollectionKey key,
+            [NotNullWhen(true)] out IGlyphTypeface? glyphTypeface)
+        {
+            glyphTypeface = null;
+
+            var stretch = (int)key.Stretch;
+
+            if (stretch < 5)
+            {
+                for (var i = 0; stretch + i < 9; i++)
+                {
+                    if (glyphTypefaces.TryGetValue(new FontCollectionKey(key.Style, key.Weight, (FontStretch)(stretch + i)), out glyphTypeface))
+                    {
+                        return true;
+                    }
+                }
+            }
+            else
+            {
+                for (var i = 0; stretch - i > 1; i++)
+                {
+                    if (glyphTypefaces.TryGetValue(new FontCollectionKey(key.Style, key.Weight, (FontStretch)(stretch - i)), out glyphTypeface))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static bool TryFindWeightFallback(
+            Dictionary<FontCollectionKey, IGlyphTypeface> glyphTypefaces,
+            FontCollectionKey key,
+            [NotNullWhen(true)] out IGlyphTypeface? typeface)
+        {
+            typeface = null;
+            var weight = (int)key.Weight;
+
+            //If the target weight given is between 400 and 500 inclusive          
+            if (weight >= 400 && weight <= 500)
+            {
+                //Look for available weights between the target and 500, in ascending order.
+                for (var i = 0; weight + i <= 500; i += 50)
+                {
+                    if (glyphTypefaces.TryGetValue(new FontCollectionKey(key.Style, (FontWeight)(weight + i), key.Stretch), out typeface))
```

Comment:
with record FontCollectionKey you can write this code:
```csharp
        var key = new FontCollectionKey(key.Style, key.Weight, key.Stretch);

        for (var i = 0; weight + i <= 500; i += 50)
            glyphTypefaces.TryGetValue(key with { Weight = (FontWeight)(weight + i)}, out typeface)
```

Reply:
This is really neat 👍